### ISSUE TITLE
Deprecate twisted.python.hashlib

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6342]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#6342|
|Type|enhancement|
|Created|2013-03-01 21:43:37Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-hashlib-6342|

[hashlib](http://docs.python.org/2.7/library/hashlib.html) was introduced in py2.5 and `twisted.python.hashlib` was added in 2007 ([#2763](https://github.com/twisted/twisted/issues/2763)) to "allow[s] application code to transparently use APIs which existed before C{hashlib} was introduced or to use C{hashlib} if it is available."

Twisted supports Python 2.6 and up and twisted's hashlib can be deprecated, and uses of it replaced, to make it easier to support py3 and discourage users from using this module.

Initially I wanted to file a ticket for porting this module to py3 but why port it when its there for (now) unsupported versions of Python.

Attachments:

* [ticket6342.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f91/f91f1f1b73365a42a7214c522a961810e4cdb30a/7b542e05a679b87c56f02a67fcbc39ad3c4c507c.patch) (18757 bytes) - added by rodney757 on 2013-04-12 17:55:37Z - 
* [ticket6342.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f91/f91f1f1b73365a42a7214c522a961810e4cdb30a/7ff930c3f6a959bff3c76e78f3b387d802fbc5ce.patch) (19453 bytes) - added by rodney757 on 2013-04-12 21:41:01Z - lastest patch
* [ticket6342.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f91/f91f1f1b73365a42a7214c522a961810e4cdb30a/e1c1f055cbf0ef439e8e240a896c3dfc2b801b54.patch) (19544 bytes) - added by rodney757 on 2013-04-16 19:30:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6342 6342
type__enhancement enhancement
reporter__thijs thijs
priority__normal normal
milestone__ 
branch__branches_deprecate_hashlib_6342 branches/deprecate-hashlib-6342
branch_author__rodney757 rodney757
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1362174217000000 1362174217000000
changetime__1368552217000000 1368552217000000
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

You can see an example deprecation in `twisted/conch/insults/__init__.py` and the applicable tests in `twisted/conch/test/test_insults.py`.

**Comment by twisted-trac** [comment]
|<img alt="rodney757's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|rodney757 set owner to rodney757<br>rodney757 set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="rodney757's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|rodney757 set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="rodney757's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|rodney757 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince set owner to rodney757|
|-|-|

Thanks for your contribution.
 1. Don't delete the existing test. Even if people shouldn't be using the code, we don't want the code to break. The tests should only be removed when the code they are testing gets removed.
   - You'll want to make sure that the warning is suppressed in those tests (using the `.suppress` attribute and `twisted.trial.util.suppress`).
 2. Similarily, the module docstring doesn't want to be completely deleted. On the other hand, it should be changed to include the deprecation information (`twisted.python.deprecate._getDeprecationDocstring` has the standard format for that). It probably also wants to be changed to indicate that twisted doesn't use it anymore.
 3. There are still a couple of uses t.p.hashlib left after applying your patch.    - `twisted/conch/client/knownhosts.py'
    - `twisted/conch/ssh/keys.py`
    - `twisted/web/server.py`
 4. Your change needs a [https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles news file].
 5. The conditional (try) logic in `t.p.hashlib` can be removed. On supported versions, the `except` will never trigger.
 6. Refer to 'stdlib' or 'standard library' not 'Python Core'.

**Comment by twisted-trac** [comment]
|<img alt="rodney757's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|rodney757 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince set owner to rodney757|
|-|-|

Thanks for your contribution.
 1. You can probably just do something like
```python
from hashlib import md5, sha1
```
    The code was structured the way it was before so it only did the import once. (Although, on further consideration, I'm not sure why the code wasn't like that to begin with.
 2. 13.0.0 was the most recent release, so 13.1.0 will be the next release. That will be the release where this will be deprecated.
 3. Calls to `.flushWarnings` should include a function to blame. This is to verify that the stacklevel argument is set correctly. It should point to the function that is invoking the deprecated functionality.
 4. The spacing between functions and classes doesn't follow the  [coding standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto5).
    - (minor) Also, there typically isn't a blank line after method docstrings
    - There is a spurious blank line at the begining of hashlib's module docstring.

**Comment by twisted-trac** [comment]
|<img alt="rodney757's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|rodney757 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [38141]) Branching to 'deprecate-hashlib-6342'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

(In [38145]) correct version number, refs [#6342](https://github.com/twisted/twisted/issues/6342)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

Thanks for your patch. The version nr was still incorrect so I fixed that and forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-hashlib-6342).

Leaving the review keyword so someone else can review (and/or merge) this since I made some minor changes to the patch (inc adding news files).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_12"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj set owner to @thijstriemstra<br>@wallrj set status to `new`|
|-|-|

## Code Review

Thanks for the branch. Here's a quick review.

Notes:
 * Branch merges cleanly to trunk
 * Tests pass locally on Fedora 18 x86_64

I spotted one minor thing:
 1. [branches/deprecate-hashlib-6342/twisted/python/test/test_hashlib.py](branches/deprecate-hashlib-6342/twisted/python/test/test_hashlib.py)
    1. test_deprecation
       1. "Test to" in docstring is unnecessary

Otherwise the branch looks fine and I couldn't find any remaining uses
of twisted.python.hashlib. Fix that docstring and then merge.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the review, I'll fix that docstring and merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

(In [38409]) address review comment, refs [#6342](https://github.com/twisted/twisted/issues/6342)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [38410]) Merge deprecate-hashlib-6342: Replace usage of twisted.python.hashlib and deprecate it in favor of the hashlib module in the standard library.

Author: rodney757
Reviewer: tom.prince, thijs, rwall
Fixes: [#6342](https://github.com/twisted/twisted/issues/6342)
